drivers/modem/modem_cellular.c

```diff
@@ -1295,6 +1296,36 @@ static void modem_cellular_run_apn_script_event_handler(struct modem_cellular_da
 	}
 }
 
+static void modem_cellular_script_failed_handler(struct modem_cellular_data *data,
+						 const struct modem_cellular_config *config)
+{
+	const uint8_t max_script_failures = 3;
+
+	data->script_failure_counter++;
+	LOG_INF("Script failed %d times", data->script_failure_counter);
+	if (data->script_failure_counter >= max_script_failures) {
+		data->script_failure_counter = 0;
+		if (data->state == MODEM_CELLULAR_STATE_CARRIER_ON) {
+			net_if_carrier_off(modem_ppp_get_iface(data->ppp));
+		}
+		modem_chat_release(&data->chat);
+		modem_cmux_release(&data->cmux);
```

Comment:
this already happens here https://github.com/zephyrproject-rtos/zephyr/blob/9da1b3cba1a2a8a448c8650fb30b38b570a9b9de/drivers/modem/modem_cellular.c#L761-L780